icommands.c

```diff
@@ -98,6 +103,137 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
   return rc;
 }
 
+/**
+ * icmd_bind - Parse 'bind' and 'macro' commands - Implements ::icommand_t
+ */
+static enum CommandResult icmd_bind(struct Buffer *buf, struct Buffer *s,
+                                    unsigned long data, struct Buffer *err)
+{
+  FILE *fpout = NULL;
+  bool valid_menu = false;
+  char tempfile[PATH_MAX];
+  struct Keymap *map = NULL, *next = NULL;
+
+  if (!MoreArgs(s))
+    mutt_buffer_strcpy(buf, "all");
+  else
+    mutt_extract_token(buf, s, 0);
+
+  if (MoreArgs(s))
+  {
+    /* More arguments potentially means the user is using the
+     * ::command_t :bind command thus we delegate the task. */
+    return MUTT_CMD_ERROR;
+  }
+
+  mutt_mktemp(tempfile, sizeof(tempfile));
+  fpout = mutt_file_fopen(tempfile, "w");
+  if (!fpout)
+  {
+    mutt_buffer_printf(err, _("Could not create temporary file %s"), tempfile);
+    return MUTT_CMD_ERROR;
+  }
+
+  const bool all_menus = mutt_str_strcasecmp(buf->data, "all") == 0;
+  for (int i = 0; i < MENU_MAX; i++)
```

Comment:
I think we should find a way to break out this block (and its children blocks) statement into functions to improve readability and make unit testing possible. The control flow is not conducive to quickly understanding what's happening.